### PR TITLE
Uniform handling of side-effects for opaque definitions

### PR DESCRIPTION
**Pull request by ppedrot**
We remove the last call to the kernel (in `Declare`) that makes a difference between opaque definitions depending on their polymorphic status. This allows to enforce statically that only opaque definition contain side-effects, leading to various code removal.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Is this a behaviour change? If so what was the reason for the previous behaviour?

**Comment by ppedrot** [comment]
This does change the behaviour when e.g. calling abstract inside a polymorphic proof. I think that the previous behaviour was mostly an oversight, due to the conflation between several phenomena at play when declaring opaque definitions, e.g. universe inference, side-effect computation and whatnot.

File: kernel/entries.ml [hunk 1]
```diff
@@ -99,11 +99,14 @@ type primitive_entry = {
 type 'a proof_output = constr Univ.in_universe_context_set * 'a
 type 'a const_entry_body = 'a proof_output Future.computation
 
-type 'a constant_entry =
-  | DefinitionEntry of definition_entry
-  | OpaqueEntry of 'a const_entry_body opaque_entry
-  | ParameterEntry of parameter_entry
-  | PrimitiveEntry of primitive_entry
+(** Dummy wrapper type discriminable from unit *)
+type 'a wrap = { wrap : 'a }
```
Comment (gares) [inline kernel/entries.ml]: Wouldn't it make sense to call this box "effects" ? Looks like unit means pure after all, while this thing is for impure stuff.
Or maybe just use an inductive with two constructors Pure != Effectful 'a.

(if it deserves a comment it means it is surprising)

Comment (ppedrot) [reply kernel/entries.ml]: This is actually temporary, I have a further path that removes this type. I had to introduce it not to have to change too much the upper layers.

**Comment by ppedrot** [comment]
@maximedenes can you review and assign? I think I have to add a CHANGES entry because it does change the observable behaviour in corner cases, but apart from this the PR is ready.

File: kernel/entries.ml (hunk 1 above)
Comment (maximedenes) [reply kernel/entries.ml]: I would still prefer that we give it a more meaningful name. Temporary things tend to stay longer than we expect...

**Review by maximedenes — COMMENTED** [review]
I'm not really competent on this side-effects stuff, but if the patch looks ok to @gares and @SkySkimmer, I can merge. Please improve the name of the wrapper type, though.

File: doc/changelog/01-kernel/10439-uniform-opaque-seff-handling.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+- Internal definitions generated by abstract-like tactics are now inlined
+  inside universe polymorphic definitions, similarly to what happens for their
```
Comment (JasonGross) [inline doc/changelog/01-kernel/10439-uniform-opaque-seff-handling.md]: Doesn't inlining only happen for Qed'd definitions (in the monomorphic case)?  Can you clarify the behavior of abstract in polymorphic definitions wrt Qed vs Defined?

**Review by JasonGross — DISMISSED** [review]

File: doc/changelog/01-kernel/10439-uniform-opaque-seff-handling.md (hunk 2 above)
Comment (ppedrot) [reply doc/changelog/01-kernel/10439-uniform-opaque-seff-handling.md]: Yes, this is missing from the description, it only happens for Qed-terminated definitions.

File: doc/changelog/01-kernel/10439-uniform-opaque-seff-handling.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+  monomorphic counterparts,
+  by Pierre-Marie Pédrot.
```
Comment (Zimmi48) [inline doc/changelog/01-kernel/10439-uniform-opaque-seff-handling.md]:
```suggestion
  (`#10439 <https://github.com/coq/coq/pull/10439>`_, by Pierre-Marie Pédrot).
```

**Comment by ppedrot** [comment]
@maximedenes What name do you prefer for `wrap`?

**Comment by maximedenes** [comment]
> What name do you prefer for `wrap`?

Anything that would be more meaningful. Just pick one.

**Comment by ppedrot** [comment]
Erm, `wrap` was already pretty meaningful to me, I wouldn't have picked that up otherwise. I can't think of a name that is not a synonymous for the function of such a type...

**Comment by SkySkimmer** [comment]
If it's general purpose then it should be in Util or some such. If it's specialized to effects then the name should indicate that.

**Comment by ppedrot** [comment]
Sure you don't just want to wait for the upcoming PR that will erase it?

**Comment by SkySkimmer** [comment]
Just call it seff_wrap and get on with it

Gaëtan Gilbert

**Comment by maximedenes** [comment]
> Sure you don't just want to wait for the upcoming PR that will erase it?

Can it be seen somewhere already?

**Comment by maximedenes** [comment]
> Just call it seff_wrap and get on with it

Yes, at least a more specialized name like that would be better.

**Comment by ppedrot** [comment]
Let's go for seff_wrap then.

**Comment by ppedrot** [comment]
Should be ready now.

**Comment by maximedenes** [comment]
@gares @SkySkimmer does this change get your approval?

**Comment by SkySkimmer** [comment]
I'm no expert on side effects, with that understanding it seems OK

**Review by gares — APPROVED** [review]
ok

**Comment by ppedrot** [comment]
@maximedenes this is ready to be merged, the CI failure is spurious.

**Comment by ppedrot** [comment]
@maximedenes ping

**Pull request merged**